lib/internal/test_runner/test.js

```diff
@@ -362,8 +362,39 @@ class Test extends AsyncResource {
   }
 
   matchesTestNamePatterns() {
-    return ArrayPrototypeSome(testNamePatterns, (re) => RegExpPrototypeExec(re, this.name) !== null) ||
-      this.parent?.matchesTestNamePatterns();
+    const matchesByNameOrParent = ArrayPrototypeSome(testNamePatterns, (re) =>
+      RegExpPrototypeExec(re, this.name) !== null ||
+      this.parent?.matchesTestNamePatterns(),
+    );
+
+    if (matchesByNameOrParent) {
+      return true;
+    }
+
+    const testNameWithAncestors = this.getTestNameWithAncestors();
+    if (!testNameWithAncestors) {
+      return false;
+    }
+
+    return ArrayPrototypeSome(testNamePatterns, (re) => RegExpPrototypeExec(re, testNameWithAncestors) !== null);
+  }
+
+  /**
+   * Returns a name of the test prefixed by name of all its ancestors in ascending order, separated by a space
+   * Ex."grandparent parent test"
+   *
+   * It's needed to match a single test with non-unique name by pattern
+   */
+  getTestNameWithAncestors() {
+    const ancestorNames = [];
+    let parent = this.parent;
+
+    for (let i = 0; i < this.nesting; i++) {
+      ArrayPrototypePush(ancestorNames, parent.name);
```

Comment:
```suggestion
      ancestorNames[this.nesting - i - 1] = parent.name;
```

Reply:
Good idea 👍🏼 

I was also thinking to use reversed loop like 
```js
for (let i = this.nesting - 1; i >= 0; i--) {
  ancestorNames[i] = parent.name;
  parent = parent.parent;
}
```
but incrementation is more "natural" 😄